Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp

```diff
@@ -0,0 +1,248 @@
+            markClientForInvalidation(client, mode);
+
+        RenderSVGResource::markForLayoutAndParentResourceInvalidation(client, needsLayout);
+    }
+
+    markAllClientLayersForInvalidation();
+}
+
+void RenderSVGResourceContainer::markAllClientLayersForInvalidation()
+{
+    if (m_clientLayers.isEmptyIgnoringNullReferences())
+        return;
+
+    auto& document = (*m_clientLayers.begin()).renderer().document();
+    if (!document.view() || document.renderTreeBeingDestroyed())
+        return;
+
+    auto inLayout = document.view()->layoutContext().isInLayout();
+    for (auto& clientLayer : m_clientLayers) {
+        // FIXME: We should not get here while in layout. See webkit.org/b/208903.
+        // Repaint should also be triggered through some other means.
+        if (inLayout) {
+            clientLayer.renderer().repaint();
+            continue;
+        }
+        if (auto* enclosingElement = clientLayer.enclosingElement())
+            enclosingElement->invalidateStyleAndLayerComposition();
+        clientLayer.renderer().repaint();
+    }
+}
+
+void RenderSVGResourceContainer::markClientForInvalidation(RenderObject& client, InvalidationMode mode)
+{
+    ASSERT(!m_clients.isEmptyIgnoringNullReferences());
+
+    switch (mode) {
+    case LayoutAndBoundariesInvalidation:
+    case BoundariesInvalidation:
+        if (is<RenderSVGShape>(client))
+            downcast<RenderSVGShape>(client).setNeedsShapeUpdate();
+        break;
+    case RepaintInvalidation:
+        if (!client.renderTreeBeingDestroyed() && !client.needsLayout())
+            client.repaint();
+        break;
+    case ParentOnlyInvalidation:
+        break;
+    }
+}
+
+void RenderSVGResourceContainer::addClient(RenderElement& client)
+{
+    m_clients.add(client);
+}
+
+void RenderSVGResourceContainer::removeClient(RenderElement& client)
+{
+    removeClientFromCache(client, false);
+    m_clients.remove(client);
+}
+
+void RenderSVGResourceContainer::addClientRenderLayer(RenderLayer& client)
+{
+    m_clientLayers.add(client);
+}
+
+void RenderSVGResourceContainer::removeClientRenderLayer(RenderLayer& client)
+{
+    m_clientLayers.remove(client);
+}
+
+void RenderSVGResourceContainer::registerResource()
+{
+    auto& treeScope = this->treeScopeForSVGReferences();
+    if (!treeScope.isIdOfPendingSVGResource(m_id)) {
+        treeScope.addSVGResource(m_id, *this);
+        return;
+    }
+
+    auto elements = copyToVectorOf<Ref<SVGElement>>(treeScope.removePendingSVGResource(m_id));
+
+    treeScope.addSVGResource(m_id, *this);
+
+    // Update cached resources of pending clients.
+    for (auto& client : elements) {
+        ASSERT(client->hasPendingResources());
+        treeScope.clearHasPendingSVGResourcesIfPossible(client);
+        auto* renderer = client->renderer();
+        if (!renderer)
+            continue;
+        SVGResourcesCache::clientStyleChanged(*renderer, StyleDifference::Layout, nullptr, renderer->style());
```

Comment:
Also very dubious -- what exactly does that do? We should avoid these question in the new resources system. IMHO we should avoid any use of SVGResourceCache::client* in LBSE, from the beginning.